doc/designs/lms.md

```diff
@@ -0,0 +1,243 @@
+signature, otherwise an error will occur.
+- Section 4 of [2] states that the same hash function MUST be used everywhere
+for OTS and LMS at all HSS levels (this means the terms 'm' and 'n' are the
+same)). ([1] only states that the hashes SHOULD be the same).
+This requires error checking when the HSS signature data is decoded.
+- Hash Functions for SHAKE and truncated SHA256 are not specified in [1],
+but are part of [2] so they should be supported.
+For truncated SHAKE256/192 support this will require the OSSL_DIGEST_PARAM_XOFLEN
+parameter to be set.
+
+Threading
+---------
+
+For HSS, multi-level trees are used for verification,
+and there is a verify process involving many hashes for each of the (1..8)
+levels, with only the final level using the input message (and the message is
+possibly streamed).
+All the other levels are one shot verify operations that must all return true in
+order to pass verification. These one-shot operations may be done in parallel
+using the new internal thread API calls (introduced for ARGON).
+When streaming only the final part of the validation involves the message,
+and this is done using a init, update and final in the main thread.
+After this the other threads must finish before verification can end.
+If threads are not supported, then a single threaded call is used instead.
+By default the thread count will be one.
+A parameter to set the threads can be done via a OSSL_PARAM.
+For one shot verification threading can still be used for the top levels.
+
+Provider Support
+----------------
+
+dupctx should not be supported by the provider since this would not apply to LMS.
+
+In order to support the new API calls there should be 2 additional dispatch
+functions..
+
+OSSL_CORE_MAKE_FUNC(int, signature_digest_verify_pbs_init,
+                    (void *ctx, const char *mdname, void *provkey,
+                     const OSSL_PARAM params[],
+                     const unsigned char *sig, size_t siglen))
+OSSL_CORE_MAKE_FUNC(int, signature_digest_verify_pbs_final,
+                    (void *ctx))
+
+
+A new HSS keymanager should support:
+
+OSSL_FUNC_KEYMGMT_NEW
+OSSL_FUNC_KEYMGMT_FREE
+OSSL_FUNC_KEYMGMT_SET_PARAMS
+OSSL_FUNC_KEYMGMT_SETTABLE_PARAMS
+OSSL_FUNC_KEYMGMT_HAS
+OSSL_FUNC_KEYMGMT_MATCH
+OSSL_FUNC_KEYMGMT_VALIDATE
+OSSL_FUNC_KEYMGMT_IMPORT
+OSSL_FUNC_KEYMGMT_IMPORT_TYPES
+OSSL_FUNC_KEYMGMT_EXPORT
+OSSL_FUNC_KEYMGMT_EXPORT_TYPES
+OSSL_FUNC_KEYMGMT_LOAD
+
+A HSS public key requires 2 OSSL_PARAM fields:
+OSSL_PKEY_PARAM_HSS_L int 
+OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY octet_string
+
+FIPS Requirements
+------------------
+
+See [FIPS 140-3 IG] "10.3.A Cryptographic Algorithm Self-Test Requirements"
+If the module implements signature verification for HSS, it shall have a CAST
+for it.
+
+Testing
+-------
+
+As well as using the test vectors supplied in [1] and [6], additional interop tests
+should be performed by generating data using other toolkits
+(such as Ciso or Bouncy Castle).
```

Comment:
Do these implement the sign side?

Reply:
Since the RFC originated from Cisco they needed to write a reference implementation. BC has the sign side. The Java SDK only implements the verify.. In my opinion it is asking for trouble implementing stateful signing (which is why it is supposed to be done in hardware only)

Reply:
There is a reference implementation available: https://github.com/cisco/hash-sigs